libbeat/publisher/queue/queue_reg.go

```diff
@@ -4,23 +4,34 @@ import (
 	"fmt"
 
 	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/feature"
 )
 
+// Namespace is the feature namespace for queue definition.
+var Namespace = "libbeat.queue"
+
 // Global queue type registry for configuring and loading a queue instance
 // via common.Config
 var queueReg = map[string]Factory{}
 
 // RegisterType registers a new queue type.
-func RegisterType(name string, f Factory) {
-	if queueReg[name] != nil {
-		panic(fmt.Errorf("queue type '%v' exists already", name))
-	}
-	queueReg[name] = f
+func RegisterType(name string, fn Factory) {
+	f := feature.New(Namespace, name, fn, feature.Undefined)
+	feature.MustRegister(f)
 }
 
 // FindFactory retrieves a queue types constructor. Returns nil if queue type is unknown
 func FindFactory(name string) Factory {
-	return queueReg[name]
+	f, err := feature.Registry.Find(Namespace, name)
+	if err != nil {
+		return nil
+	}
+	factory, ok := f.Factory().(Factory)
+	if !ok {
+		return nil
+	}
+
+	return factory
```

Comment:
Using reflection, this pattern can be simplified to:

```
func FindFactory(name string) (f Factory) {
  if err := feature.Registry.Lookup(Namespace, name, &f); err != nil {
    // log error?
  }
  return
}
```

Using reflaction based function calling support, we could even implement an 'Instantiate/Invoke' method in the registry (maybe too much magic, though).

Reply:
I think I would prefer to stay away from reflexion as much as possible in the context of plugin, this force user to make the choice.

Reply:
I've created a private bundleable interface implement by both Feature and Bundle.

I've kept the `Feature` and `Bundle`  type public, since we will interact with it for creating new Features and filtering / merging bundle.